From dce6dd223c24a75cecaee41bec78141e887282f7 Mon Sep 17 00:00:00 2001
From: Paul Drews <paul.drews@intel.com>
Date: Tue, 22 Mar 2011 01:35:00 -0700
Subject: init: Make control-C work on the console

Job control (e.g., control-C) doesn't work for the Android
console shell.  This change sets the controlling tty for the
console service so that job control works.

Design:  When a service is launched with the "console"
attribute, issue a TIOCSCTTY ioctl to its stdin
file-descriptor to set it as the controlling tty.  The ioctl
is issued with the "try" instead of "steal" option in case
there is already ownership of the controlling tty.
Presumably only the console shell has the "console"
attribute, so this should not be a concern in any case.

Change-Id: I281c1beb0ae1d4ba17dc337a2c753459c0bd217b
Signed-off-by: Paul Drews <paul.drews@intel.com>
---
 init/init.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/init/init.c b/init/init.c
index 91ac7b9..cd302e6 100755
--- a/init/init.c
+++ b/init/init.c
@@ -277,6 +277,8 @@ void service_start(struct service *svc, const char *dynamic_args)
         if (needs_console) {
             setsid();
             open_console();
+            /* Try to take, but don't steal, controlling tty */
+            ioctl(0, TIOCSCTTY, 0);
         } else {
             zap_stdio();
         }
-- 
1.7.4.1

